### PR TITLE
Add option to hash BrowserID usernames

### PR DESCRIPTION
**Pull request by thedod**
If there's a config option in the browserid called hash_secret, usernames become a hash of the BrowserID email, with hash_secret as salt. If hash_secret doesn't exist - it works just like before (username is the BrowserID email address).

### CODE REVIEW

**Comment by jhs** [comment]
Awesome. Would it be possible for you to modify the style slightly, either by editing the history and submitting another push request, or even just another commit with the syntax fixes?

I use an unconventional style but I suppose at least we can keep it consistent.

The basic idea is that Erlang punctuation really _separates_ two expressions, it doesn't so much _terminate_ one. Especially the comma, in my head I read it as "statement 1... _and then_ statement 2... _and then_ statement 3... _and that's all_." Finally, you may notice that case statements have a very Python-like indentation which I find clarifies the code.

Therefore, I am trying an experiment: place the punctuation in the front where it is both easier to see and also easier to ignore. (Also diffs are better since one needn't change the previous line so often.) One catch is that the first statement in a function must not have a comma, that is why I have a no-op `ok` expression.

```
some_function() -> ok
    , first_statement()
    , second_statement
    , case Whatever
        of first_case -> ok
            , foo
            , bar
        ; second_case -> ok
            , baz()
        end
    , final_expression()
    .
```

File: README.md [hunk 1]
```diff
@@ -4,9 +4,9 @@ This is a plugin for CouchDB to support Mozilla's BrowserID standard.
 
 ## So Simple
 
-First, for security reasons, BrowserID is disabled by default. Go to Futon's *Configuration* section. Look for *BrowserID* and change *enabled* to `true`.
+First, for security reasons, BrowserID is disabled by default. Go to Futon's *Configuration* page. Look for the *browserid* section, and change the *enabled* option to `true`.
 
-Good. Put this in your Couch app.
+Good. Put this in your Couch app.
```
Comment (jhs) [inline README.md]: Would you mind removing this whitespace please? It's a splinter in my mind. :)

File: src/couch_httpd_browserid.erl [hunk 2]
```diff
@@ -126,8 +143,16 @@ verify_id_with_crutch(VerifyURL, Assertion, Audience) ->
                 , Email  = couch_util:get_value(<<"email">> , Resp, undefined)
                 , case {Status, Email}
                     of {<<"okay">>, Email} when Email =/= undefined -> ok
-                        , ?LOG_DEBUG("BrowserID verified: ~p", [Resp])
-                        , {ok, Resp}
+                        , Newresp = [
```
Comment (jhs) [inline src/couch_httpd_browserid.erl]: Would you mind changing the syntax to a more explicit punctuation-first style? Sorry to nit-pick, I can't help myself! But I think in the long term, consistent style (even if bizarre) is less distracting.

In this situation, I like the NodeJS punctuation-first style, except the Erlang 4-space indentation.

```
, Newresp =
    [ {<<"status">>, <<"okay">>}
    , {<<"email">>, hash_if_required(Email)}
    , etc, etc, etc
    ]
, ?LOG_DEBUG("Browser ID verified: ~p", [Newresp])
```

In fact, either from my old Perl days or my old Emacs days, I really prefer to trim that by one line of code and keep it right inline with the original statement! And furthermore, I like to align the commas (or in Javascript, the colon) to achieve a highly tabular format.

```
Example = [ {<<"number">>   , 1234.567}
          , {<<"a string">> , <<"This is a string value">>}
          , {<<"an object">>, {[ {<<"sub_num">>, -12.34}
                               , {<<"str">>    , <<"Hello!">>}
                               , {<<"list">>   , [small, inline, lists, are, ok]}
                               ]}
          ]
```

However I unfortunately break this rule if the lines would wrap on an editor. I believe 80 columns is far too conservative considering modern displays but one must draw the line somewhere! (On my 11-inch screen, Vim gives me 168 columns of width.)

In short, I have not come to a consistent conclusion about when to use which style, but I always use one or the other. I believe programmers are losing some readability and clarity (which is extremely important with crypto and authentication code!) when they don't make as clear as possible the vertical relationship between lines of code.

File: src/couch_httpd_browserid.erl [hunk 3]
```diff
@@ -23,6 +23,23 @@
...
+% If missing or empty (i.e. admin doesn't want to hash usernames), simply returns the argument
+% (this is the default)
+% If you use hash_secret, make sure the string is long enough and cryptographically random
+% Tip: use one of the strings from https://api.wordpress.org/secret-key/1.1/ :)
+
+hash_if_required(Email) ->
+    Hashkey = couch_config:get("browserid", "hash_secret", undefined)
+    , case Hashkey
+        of undefined -> Email
+        ; Key -> <<Mac:160/integer>> = crypto:sha_mac(?l2b(Key),Email)
```
Comment (jhs) [inline src/couch_httpd_browserid.erl]: See the problem? This is why I decided to have the initial no-op `ok`. It looks funny but my hope is by keeping a very clean vertical alignment, it improves readability and understandability and ultimately prevents bugs.

```
; Key -> ok
    , <<Mac:160/integer>> = crypto:sha_mac(?l2b(Key), Email)
```

**Comment by thedod** [comment]
I believe I've fixed everything.
If you don't want the ugly commit history, I can simply delete my fork, for again and make it look like I was _born_ smart :)
What do you say?

**Comment by thedod** [comment]
OK.
I'll do the switcheroo
